### PR TITLE
Ticket 4: Login API

### PR DESCRIPTION
**Pull request by KanyaLee**

### CODE REVIEW

File: jest.config.js
Comment (chenyou-H) [inline jest.config.js:1]: should not delete an existing file

File: pages/api/login.ts
Comment (chenyou-H) [inline pages/api/login.ts:9]: This login API handler will respond to all kinds of HTTP requests like GET, DELETE, and so on. Your login API should only process POST requests and not respond to other requests. Please set some conditions for your response

File: pages/api/login.ts
Comment (chenyou-H) [inline pages/api/login.ts:32]: can you leave an example for JWT_SECRET in the env. sample? I can't test it without knowing what it is.

File: models/User.ts
Comment (chenyou-H) [inline models/User.ts]: can you update the User schema in order to avoid the merge conflict in the future? 


```javascript
const UserSchema = new mongoose.Schema({
  firstname: {
    type: String,
    required: [true, 'Please provide a first name for this user']
  },
  lastname: {
    type: String,
    required: [true, 'Please provide a last name for this user']
  },
  email: {
    type: String,
    required: [true, 'Please provide a email for this user'],
    unique: true,
    trim: true,
    lowercase: true
  },
  password: {
    type: String,
    required: [true, 'Please provide a password for this user']
  },
}, {
  timestamps: true, // this will add createdAt and updatedAt timestamps
});
```

File: pages/api/login.test.ts
Comment (chenyou-H) [inline pages/api/login.test.ts]: can you use Mongoose instead of the MongoDB driver and use the test database we already had to test the API?

**Review by PHANTOMS0308 — CHANGES_REQUESTED** [review]
Generally great, however, there are a few things need to be fixed before we can merge.

- First of all, you made too many changes to the initial configuration. Usually, you should not delete or modify any configuration before you discussing it with the team leads and teammates. Because any change you made in shared files might lead to serious bug or error for other developers. But I understand that our initial configuration did not considering the API testing, therefore, if you look at the latest pull request at #18 , providing an alternative solution (waiting for code review and developers experiments) to achieve the problem you were trying to solve by changing many configuration.
- Secondly, for all test that connects database, make sure to disconnect before the test end.
- Meanwhile, I would like you to fix the pull request after @chenyou-H, because Alex defined schema and functions for you and him already, I would like you to update the latest dev branch after he merged, so that you can integrated with his schema and functions. 

If you have any concern or problems, DM me or Alex, we will be more than willing to assist you.

File: models/User.ts
Comment (PHANTOMS0308) [inline models/User.ts]: Good job, this is a good use of middleware here. However, we prefer the use of separate functions to achieve the business logic. You could check out #15 for better understanding of what I mean. Also you don't need to define such functions here, since Alex already did. (But I understand you needed those to proceed for testing, so I suggest you wait till Alex pull then based your change on his)

File: models/User.ts
Comment (PHANTOMS0308) [inline models/User.ts]: Same thing over here, a separate function will be better, however, this comparePassword is also defined by Alex

File: models/User.ts
Comment (PHANTOMS0308) [inline models/User.ts:1]: Actually, after Alex merged, you don't even need this file

File: package.json
Comment (PHANTOMS0308) [inline package.json:1]: Make sure to revert it back to the original one, besides `jsonwebtoken`, there should be no other dependency added

File: pages/api/login.test.ts
Comment (PHANTOMS0308) [inline pages/api/login.test.ts]: Once my #18 were approved and merged, you could `dbConnect()` in the `beforeAll` and `dbDisconnect()` in the `afterAll`

Comment (PHANTOMS0308) [reply pages/api/login.test.ts]: You may also have a logic to create a new collection called `users-test-login` before all test runs and delete that collection after it. The reason is that you would like to have a clean collection to feed your data. What if you encounter a user whose email is exactly the same as your test user email, but password is different, then many of your test will fail. 

If you choose not to create a new collection, then you need an extra logic to clear the collection you are testing with

File: pages/api/login.test.ts
Comment (PHANTOMS0308) [inline pages/api/login.test.ts]: Put that right after `beforeEach` for better code readability.

File: pages/api/login.test.ts
Comment (PHANTOMS0308) [inline pages/api/login.test.ts]: You could now use `node-mocks-http`, since all bugs have been fixed. Using that library will make testing easier

File: pages/api/login.test.ts
Comment (PHANTOMS0308) [inline pages/api/login.test.ts]: But using Jest mock here is pretty smart, without the help of `node-mocks-http`

File: pages/api/login.test.ts
Comment (PHANTOMS0308) [inline pages/api/login.test.ts]: First of all, always make sure your unit/integration tests shall all be passing when you make a pull request.

The reason why this test case failed is that you put `password: 'validpassword'`. Since the password stored in the database was hashed, then you should provide a hashed password. And that valid hashed password should be stored and updated by `beforeEach`. Currently you don't have the access to that valid hashed password, therefore you need to fix this.

File: pages/api/login.test.ts
Comment (PHANTOMS0308) [inline pages/api/login.test.ts]: For user login, there are mainly three scenarios.
1. User exist, password correct
2. User exist, password wrong
3. User not exist, password correct/wrong

You test about the last two, but not the first one.

File: pages/api/login.test.ts
Comment (PHANTOMS0308) [inline pages/api/login.test.ts:1]: Follow what Alex did, put the test inside `@/__test__/api/`.

File: pages/api/login.ts
Comment (PHANTOMS0308) [inline pages/api/login.ts]: Again, use Alex's defined functions, instead of creating such functions inside your page

File: pages/api/login.ts
Comment (PHANTOMS0308) [inline pages/api/login.ts]: Talk with @chenyou-H , all backend developers should make sure two things
1. All your models functions should return Data or throw Error in consistent format
2. All your API endpoint should return Data or Error object(as json) in consistent format

File: pages/api/login.ts
Comment (PHANTOMS0308) [reply pages/api/login.ts:32]: Also the logic of creating a JWT token after user login, is better to be be extracted as a util functions.

File: pages/api/login.ts
Comment (PHANTOMS0308) [inline pages/api/login.ts]: Token should be sent back as a cookie inside the response header with two configuration `HttpOnly` and `SameSite`. If you are not very familiar with the cybersecurity aspect, feel free to talk with me

File: .babelrc
Comment (zhiyuan-Z) [inline .babelrc:1]: Shouldn't be necessary.

File: pages/api/login.test.ts
Comment (zhiyuan-Z) [reply pages/api/login.test.ts]: Actually, bcrypt.compare() compares a plain password with a hashed one. It probably fails because of some other reasons.

**Review by zhiyuan-Z — COMMENTED** [review]
Made a few changes to the configuration, testing and login api.

File: pages/api/login.ts
Comment (PHANTOMS0308) [inline pages/api/login.ts:49]: Perfect and Secure usage of JWT in cookies!

**Review by PHANTOMS0308 — APPROVED** [review]
Well done on the refactoring. However it sort of reminding me that we did not have any JWT in the response header after user first sign up. Anyway, the login part is well done. We have sufficient time to refactor others in the future, right now we wanna merge this logic and Vijay and Zack can refer to the JWT logic in this codebase

**Review by chenyou-H — APPROVED** [review]
You are awesome for making these updates. LGTM!

**Pull request merged**
